### PR TITLE
drivers/pn532: adapted to SPI interface changes

### PR DESCRIPTION
**Pull request by haukepetersen**
missed in #4780: driver needed adaption to SPI interface changes for its SPI mode.

### CODE REVIEW

**Comment by haukepetersen** [comment]
build the test-app with PN532_MODE=spi BOARD=xxx make ...

File: drivers/pn532/pn532.c [hunk 1]
```diff
@@ -120,23 +124,21 @@ int pn532_init(pn532_t *dev, const pn532_params_t *params, pn532_mode_t mode)
...
-        ret = i2c_init_master(dev->conf->i2c, I2C_SPEED_FAST);
+        if (i2c_init_master(dev->conf->i2c, I2C_SPEED_FAST) != 0) {
+            DEBUG("pn532: initialization of I2C bus failed\n");
+            return -1;
+        }
 #endif
     }
     else {
 #ifdef PN532_SUPPORT_SPI
-        ret = spi_init_master(dev->conf->spi, SPI_CONF_FIRST_RISING,
```
Comment (cgundogan) [inline drivers/pn532/pn532.c]: `ret` is initially set to `-1` and this function returns `ret` => `pn532_init` always returns `-1` this way and thusly fails

**Comment by cgundogan** [comment]
cf. https://github.com/haukepetersen/RIOT/pull/36 some fixes to get the `pn532` driver working again (SPI_MODE)

File: drivers/pn532/pn532.c (hunk 1 above)
Comment (haukepetersen) [inline drivers/pn532/pn532.c]: ups...

**Comment by haukepetersen** [comment]
fixed init function and removed debug output from test makefile

**Comment by cgundogan** [comment]
@haukepetersen please look at https://github.com/haukepetersen/RIOT/pull/36 there is another bug

**Comment by cgundogan** [comment]
ACK works as expected now. Please squash 👍

**Comment by haukepetersen** [comment]
squashed.

File: drivers/pn532/pn532.c [hunk 2]
```diff
@@ -213,17 +214,16 @@ static int _read(pn532_t *dev, char *buff, unsigned len)
...
         gpio_set(dev->conf->nss);
         spi_release(dev->conf->spi);
-        if (ret >= 0) {
-            buff[0] = 0x80;
-            reverse(buff, ret);
-            ret += 1;
-        }
+
+        buff[0] = 0x80;
+        reverse(buff, ret);
```
Comment (cgundogan) [inline drivers/pn532/pn532.c]: no idea what happened, but this fix didn't survive the squash operation (: please replace `ret` with `len`.

File: tests/driver_pn532/main.c [hunk 3]
```diff
@@ -44,8 +44,15 @@ int main(void)
...
-
-    int ret = pn532_init_i2c(&pn532, &pn532_conf[0]);
+    int ret;
+
+#if defined(PN532_SUPPORT_I2C)
+    ret = pn532_init_i2c(&pn532, &pn532_conf[0]);
+#elif defined(PN532_SUPPORT_SPI)
+    ret = pn532_init_spi(&pn532, &pn532_conf[0]);
+#else
+  #error "Error: bus mode not defined"
```
Comment (lebrush) [inline tests/driver_pn532/main.c]: This check is already included in `pn532.h` :-)

File: drivers/pn532/pn532.c (hunk 2 above)
Comment (haukepetersen) [inline drivers/pn532/pn532.c]: should be ok now.

File: drivers/pn532/pn532.c [hunk 4]
```diff
@@ -181,14 +181,15 @@ static int _write(pn532_t *dev, char *buff, unsigned len)
     }
     else {
 #ifdef PN532_SUPPORT_SPI
-        spi_acquire(dev->conf->spi);
+        spi_acquire(dev->conf->spi, GPIO_UNDEF, SPI_MODE, SPI_CLK);
```
Comment (lebrush) [inline drivers/pn532/pn532.c]: Documentation in `spi.h` says cs shall be set to `SPI_CS_UNDEF` instead of `GPIO_UNDEF`

Comment (lebrush) [inline drivers/pn532/pn532.c]: (also in the other places)

**Comment by lebrush** [comment]
Btw, I really like that the test application now supports spi too :-)

File: tests/driver_pn532/main.c (hunk 3 above)
Comment (haukepetersen) [inline tests/driver_pn532/main.c]: true

**Comment by haukepetersen** [comment]
Comments are addressed.

> (also in the other places)

you are absolutely right, will propose a fix.

**Comment by lebrush** [comment]
Nope codewise looks fine :-) but now some boards are not happy about SPI... maybe make i2c or spi the default one for the test application to solve this

```
/data/riotbuild/drivers/include/pn532.h:33:2: error: #error Please define PN532_SUPPORT_I2C and/or PN532_SUPPORT_SPI to enable the functionality on this device
 #error Please define PN532_SUPPORT_I2C and/or PN532_SUPPORT_SPI to enable \
  ^~~~~
/data/riotbuild/Makefile.base:64: recipe for target '/data/riotbuild/tests/driver_pn532/bin/calliope-mini/tests_driver_pn532/main.o' failed
make[1]: *** [/data/riotbuild/tests/driver_pn532/bin/calliope-mini/tests_driver_pn532/main.o] Error 1
make[1]: *** Waiting for unfinished jobs....
In file included from /data/riotbuild/drivers/pn532/pn532.c:26:0:
/data/riotbuild/drivers/include/pn532.h:33:2: error: #error Please define PN532_SUPPORT_I2C and/or PN532_SUPPORT_SPI to enable the functionality on this device
 #error Please define PN532_SUPPORT_I2C and/or PN532_SUPPORT_SPI to enable \
  ^~~~~
/data/riotbuild/Makefile.base:64: recipe for target '/data/riotbuild/tests/driver_pn532/bin/calliope-mini/pn532/pn532.o' failed
make[3]: *** [/data/riotbuild/tests/driver_pn532/bin/calliope-mini/pn532/pn532.o] Error 1
/data/riotbuild/Makefile.base:20: recipe for target 'ALL--/data/riotbuild/drivers/pn532' failed
make[2]: *** [ALL--/data/riotbuild/drivers/pn532] Error 2
/data/riotbuild/Makefile.base:20: recipe for target 'ALL--/data/riotbuild/drivers' failed
make[1]: *** [ALL--/data/riotbuild/drivers] Error 2
/data/riotbuild/Makefile.include:275: recipe for target 'all' failed
make: *** [all] Error 2
```

File: tests/driver_pn532/Makefile [hunk 5]
```diff
@@ -11,11 +11,26 @@ TEST_PN532_I2C ?= I2C_DEV\(0\)
...
 CFLAGS += -DTEST_PN532_I2C=$(TEST_PN532_I2C)
 CFLAGS += -DTEST_PN532_RESET=$(TEST_PN532_RESET)
 CFLAGS += -DTEST_PN532_IRQ=$(TEST_PN532_IRQ)
-CFLAGS += -DPN532_SUPPORT_I2C
+CFLAGS += -DTEST_PN532_SPI=$(TEST_PN532_SPI)
+CFLAGS += -DTEST_PN532_NSS=$(TEST_PN532_NSS)
+
+# select if you want to build the SPI or the I2C version of the driver:
+# set PN532_MODE to `i2c` or to `spi`
+TPN532_MODE ?= i2c
```
Comment (cgundogan) [inline tests/driver_pn532/Makefile]: s/`TPN532_MODE`/`PN532_MODE`

**Comment by haukepetersen** [comment]
hm, that capital `T` must have accidentally slipped in. Buildtest now runs fine again locally...

**Comment by cgundogan** [comment]
tested again with my local setup. looks fine and both CIs agree => GO

**Comment by PeterKietzmann** [comment]
@lebrush can you please provide a backport to the 2017.01 release branch?

**Comment by lebrush** [comment]
:+1: will do

**Pull request merged**